internal/pkg/utils/url/url.go

```diff
@@ -23,30 +25,93 @@ func ParseQuery(query string) (m *orderedmap.OrderedMap, err error) {
 			continue
 		}
 		key, value, _ := strings.Cut(key, "=")
-		key, err1 := url.QueryUnescape(key)
-		if err1 != nil {
-			if err == nil {
-				err = err1
+		key, err = url.QueryUnescape(key)
+		if err != nil {
+			return m, err
+		}
+		value, err = url.QueryUnescape(value)
+		if err != nil {
+			return m, err
+		}
+
+		path, sliceAppend, err := parseKey(key)
+		if err != nil {
+			return m, err
+		}
+
+		if sliceAppend {
+			existingValue, found, _ := m.GetNestedPath(path)
+			if !found {
+				err = m.SetNestedPath(path, []any{value})
+				if err != nil {
+					return m, err
+				}
+			} else {
+				if existingValueSlice, ok := existingValue.([]any); ok {
+					err = m.SetNestedPath(path, append(existingValueSlice, value))
+					if err != nil {
+						return m, err
+					}
+				} else {
+					err = errors.Errorf("invalid square brackets in query")
+					return m, err
+				}
 			}
 			continue
 		}
-		value, err1 = url.QueryUnescape(value)
-		if err1 != nil {
-			if err == nil {
-				err = err1
-			}
+
+		err = m.SetNestedPath(path, value)
+		if err != nil {
+			return m, err
+		}
+	}
+	return m, err
+}
+
+// parseKey transforms a given url key string to orderedmap.Path.
+// "key[subkey]" would give the same result as orderedmap.PathFromStr("key.subkey")
+// "key[123]" would give the same result as orderedmap.PathFromStr("key[123]")
+// [] at the end of the string will cause sliceAppend = true.
+func parseKey(str string) (path orderedmap.Path, sliceAppend bool, err error) {
+	parts := strings.FieldsFunc(str, func(r rune) bool {
+		return r == '['
+	})
+
+	for i, part := range parts {
+		if i == 0 {
+			path = append(path, orderedmap.MapStep(part))
 			continue
 		}
-		existingValue, found := m.Get(key)
-		if found {
-			if existingValueSlice, ok := existingValue.([]any); ok {
-				m.Set(key, append(existingValueSlice, value))
+		if part[len(part)-1] != ']' {
+			err = errors.Errorf(`Unable to parse key "%s"`, str)
+			return path, sliceAppend, err
+		}
+		part = part[:len(part)-1]
+		if part == "" {
+			if i == len(parts)-1 {
+				sliceAppend = true
 			} else {
-				m.Set(key, []any{existingValue, value})
+				err = errors.Errorf(`Unable to parse key "%s"`, str)
 			}
+			return path, sliceAppend, err
+		}
+
+		if isNumeric(part) {
```

Comment:
```
if  v, err := strconv.ParseInt(part, 10, 64); err != nil` {
  path = append(path, orderedmap.MapStep(part))
} else {
  path = append(path, orderedmap.SliceStep(v))
}
```
No need to check for numeric and casting it using `Atoi`. I use `ParseInt` but consider it as equal.